### PR TITLE
FIX: supplier order dispatch: order not automatically classified if containing services or free lines

### PR DESCRIPTION
**Pull request by marc-dll**

### CODE REVIEW

File: htdocs/fourn/class/fournisseur.commande.class.php [hunk 1]
```diff
@@ -3177,9 +3177,26 @@ public function calcAndSetStatusDispatch(User $user, $closeopenorder = 1, $comme
 
     				// Build array with quantity deliverd by product
     				foreach ($supplierorderdispatch->lines as $line) {
+    				    if (empty($line->fk_product)) {
+    				        continue;
+    				    }
+
+    				    if (empty($conf->global->STOCK_SUPPORTS_SERVICES)) {
```
Comment (eldy) [inline htdocs/fourn/class/fournisseur.commande.class.php:3184]: Option STOCK_SUPPORTS_SERVICES should have effect on stock quantity but not on the fact that we can deliver or receive a product of type service into an order.
We should be able to make a reception or delivery of service (this is not related to stock, only stock increase/decrease is related to stock)

Comment (marc-dll) [reply htdocs/fourn/class/fournisseur.commande.class.php:3184]: Then, if I follow what you say, the correct fix is to display service lines in the "Item receipts" tab of supplier orders whatever the setting of STOCK_SUPPORTS_SERVICES is, because right now, service lines are hidden if STOCK_SUPPORTS_SERVICES is empty.

Would it be OK this way ?

Comment (eldy) [reply htdocs/fourn/class/fournisseur.commande.class.php:3184]: It seems the same var is used for the 2 purposes. We must choose a direction. So i took a look at customer  side and shipment.
It seems that for customer, STOCK_SUPPORTS_SERVICES = 1 means:
- it affect stock of service
- we can include services into shipment but the remain to ship is always 0. So we don't manage reception for services. We will have to introduce another value for this, for example = 2 may means, we allow also reception or shipment of services.

So ignore my comment and we can follow the direction where STOCK_SUPPORTS_SERVICES = 1 is no stock change + no reception/sending, even if line is visible into shipment/document. Can you confirm your initial PR go into this direction ?

Comment (marc-dll) [reply htdocs/fourn/class/fournisseur.commande.class.php:3184]: I just tested the customer side in 11.0 and if i'm not mistaken, it's doesn't seem to match what you say as it is completely straightforward :

* if empty(STOCK_SUPPORTS_SERVICES) : services can't be part of shipments, therefore, there are no way stock can be decremented for services lines when shipment status triggers the stock calculation. It is also like thant when triggered by order or invoice status change.
* else : services can be included in shipments, stock is correctly decremented whether the change is configured to be done at shipment validation or closing (and again, same when configured from order or invoice)

It's the same in the supplier side, except that if the order contains a free line, or if it contains a service and STOCK_SUPPORTS_SERVICES is empty(), the order can never be automatically classified because those lines don't appear in the dispatch tab while stille being counted as not delivered. The PR excludes those lines from the count when necessary. This particular behaviour is correct in the customer side : when the shipment is closed, the order is correctly set shipped.

I could be wrong, but as far as I understand the whole issue, separating configuration of stock movements from shipment/reception/dispatch line inclusion is unrelated to the bug I am proposing a fix for. I'm OK with it though ;-)

**Comment by eldy** [comment]
PR seems to have been addressed by 5bae9db0a82292d2eb2c01268cce2b738bc488e3, so i close this one.
